pkg/util/objectpool/objectpool.go

```diff
@@ -0,0 +1,192 @@
+		store        chan IPoolObject            // store the object
+		checkWhenGet bool                        // whether to health check when get IPoolObject
+		checkWhenPut bool                        // whether to health check when put IPoolObject
+	}
+
+	// Spec Pool's spec
+	Spec struct {
+		InitSize     int32                       // initial size
+		MaxSize      int32                       // max size
+		New          func() (IPoolObject, error) // create a new object
+		CheckWhenGet bool                        // whether to health check when get IPoolObject
+		CheckWhenPut bool                        // whether to health check when put IPoolObject
+	}
+)
+
+// New returns a new pool
+func New(initSize, maxSize int32, new func() (IPoolObject, error)) *Pool {
+	return NewWithSpec(Spec{
+		InitSize:     initSize,
+		MaxSize:      maxSize,
+		New:          new,
+		CheckWhenGet: true,
+		CheckWhenPut: true,
+	})
+}
+
+// NewWithSpec returns a new pool
+func NewWithSpec(spec Spec) *Pool {
+	p := &Pool{
+		initSize:     spec.InitSize,
+		maxSize:      spec.MaxSize,
+		new:          spec.New,
+		checkWhenPut: spec.CheckWhenPut,
+		checkWhenGet: spec.CheckWhenGet,
+		store:        make(chan IPoolObject, spec.MaxSize),
+	}
+	if err := p.init(); err != nil {
+		logger.Errorf("new pool failed %v", err)
+		return nil
+	}
+	return p
+}
+
+// Validate validate
+func (s *Spec) Validate() error {
+	if s.InitSize > s.MaxSize {
+		s.MaxSize = s.InitSize
+	}
+	if s.MaxSize <= 0 {
+		return fmt.Errorf("pool max size must be positive")
+	}
+	if s.InitSize < 0 {
+		return fmt.Errorf("pool init size must greate than or equals 0")
+	}
+
+	return nil
+}
+
+// initializes the pool with initSize of objects
+func (p *Pool) init() error {
+	for i := 0; i < int(p.initSize); i++ {
+		iPoolObject, err := p.new()
+		if err != nil || !iPoolObject.HealthCheck() {
+			logger.Errorf("create pool object failed or object of pool is not healthy when init pool:, err %v", err)
+			continue
+		}
+		p.size++
+		p.store <- iPoolObject
+	}
+	return nil
+}
+
+// Get returns an object from the pool,
+// if no free object, it will create a new one if the pool is not full
+// if the pool is full, it will block until an object is returned to the pool
+func (p *Pool) Get(ctx context.Context) (IPoolObject, error) {
+	backoff := 1
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case poolObject := <-p.store:
+			if p.checkWhenGet && !poolObject.HealthCheck() {
+				p.destroyIPoolObject(poolObject)
+				if poolObject = p.createIPoolObject(); poolObject != nil {
+					return poolObject, nil
+				}
+				continue
+			}
+			return poolObject, nil
+		default:
+			if iPoolObject := p.createIPoolObject(); iPoolObject != nil {
+				return iPoolObject, nil
+			}
+			for i := 0; i < backoff; i++ {
+				runtime.Gosched()
+			}
+			if backoff < maxBackoff {
+				backoff <<= 1
+			}
+		}
```

Comment:
this default case could still consume a lot of CPU, I think we can leverage `sync.Cond` to avoid it (but this may involve some code changes).